torch/csrc/jit/script/compiler.cpp

```diff
@@ -72,14 +176,106 @@ struct to_ir {
       }
     }
   }
+
+  std::shared_ptr<Environment> emitSingleIfBranch(
+      Block* b,
+      const List<TreeRef> branch,
+      std::unordered_set<std::string>* mutated_parent_values) {
+    environment_stack = std::make_shared<Environment>(b, environment_stack);
+    WithInsertPoint guard(*def.graph, b);
+    emitStatements(branch);
+
+    for (const auto& kv : environment_stack->value_table) {
+      if (environment_stack->findInParentFrame(kv.first)) {
+        mutated_parent_values->insert(kv.first);
+      }
+    }
+    auto save_env = environment_stack;
+    environment_stack = environment_stack->next;
+    return save_env;
+  }
+
   void emitIf(const If& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(Symbol("If"), 0));
+    n->addInput(cond_value);
+    auto* true_block = n->addBlock();
+    auto* false_block = n->addBlock();
+
+    // Emit both blocks once to get the union of all mutated values
+    std::unordered_set<std::string> mutated_parent_values;
+    std::shared_ptr<Environment> save_true, save_false;
+    save_true = emitSingleIfBranch(
+        true_block, static_cast<List<TreeRef>>(stmt.trueBranch()), &mutated_parent_values);
+    save_false = emitSingleIfBranch(
+        false_block, static_cast<List<TreeRef>>(stmt.falseBranch()), &mutated_parent_values);
+
+    std::vector<std::string> sorted_mutations(
+        mutated_parent_values.begin(), mutated_parent_values.end());
+    std::sort(sorted_mutations.begin(), sorted_mutations.end());
+
+    // Register outputs in each block
+    environment_stack = save_true;
+    for (const auto& x : sorted_mutations) {
+      true_block->registerOutput(environment_stack->getVar(x, stmt));
+    }
+    environment_stack = save_false;
+    for (const auto& x : sorted_mutations) {
+      false_block->registerOutput(environment_stack->getVar(x, stmt));
+    }
+    environment_stack = environment_stack->next;
+
+    // Add op outputs
+    for (const auto& x : sorted_mutations) {
+      environment_stack->setVar(x, n->addOutput());
+    }
   }
 
   void emitWhile(const While& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    // TODO: scan outputs
```

Comment:
The scan outputs spec in ONNX still seems pretty unclear to me.

1. There's no example
2. What are the dimensionality requirements for `scan_outputs`? (Most stringent: they have the same exact same size. Anything weaker?)

I suppose a very clear way to define semantics is to give the algorithm for unrolling at any value of N.